Flatten cloned images of snapshot during snapshot deletion

kasabe28 opened the pull request:
# Proposed Changes

- Flatten cloned images of snapshot while deleting snapshot
- Delete volume only if it has no snapshot.
- Delete volume os-image only if it is not referenced by any other volumes
- Improve error handling in `CreateVolume()`

Fixes #809

kasabe28 commented on internal/controllers/snapshot_controller.go:
```diff
@@ -169,8 +169,22 @@ func (r *SnapshotReconciler) removeSnapshot(log logr.Logger, snapshotID string,
+		}
+		defer closeImage(log, img)
+
+		log.V(2).Info("Flatten cloned image", "ClonedImageID", clonedImgName)
```
We don't have snapshot event recorder. Maybe we can add event recorder and record all snapshot events in a separate PR.

balpert89 commented on internal/controllers/snapshot_controller.go:
```diff
@@ -169,8 +169,22 @@ func (r *SnapshotReconciler) removeSnapshot(log logr.Logger, snapshotID string,
+		defer closeImage(log, img)
+
+		log.V(2).Info("Flatten cloned image", "ClonedImageID", clonedImgName)
+		if err := img.Flatten(); err != nil {
```
As flattening an image adds considerable time consumption we may want to limit parallel flattening operations.

sujeet01 commented on internal/controllers/snapshot_controller.go:
```diff
@@ -169,8 +169,22 @@ func (r *SnapshotReconciler) removeSnapshot(log logr.Logger, snapshotID string,
-		return fmt.Errorf("unable to delete snapshot: still in use")
+	for _, clonedImgName := range imgs {
+		log.V(2).Info("Open cloned image to flatten", "ClonedImageID", clonedImgName)
+		img, err := librbd.OpenImage(ioCtx, clonedImgName, librbd.NoSnapshot)
```
`img` in the loop here shadows the above function parameter `img` variable. Please use  different variable name in the loop.

```suggestion
		clonedImg, err := librbd.OpenImage(ioCtx, clonedImgName, librbd.NoSnapshot) 
```

sujeet01 commented on internal/controllers/snapshot_controller.go:
```diff
@@ -169,8 +169,22 @@ func (r *SnapshotReconciler) removeSnapshot(log logr.Logger, snapshotID string,
+			}
+			continue
+		}
+		defer closeImage(log, img)
```
We should not use `defer` in the for loop, since `defer` does not get executed until the function returns. This will keep many images open if we have multiple clones, causing resource exhaustion. 
In the for loop here, we need to explicitly close the image after use (after flattening). 

Better approach would be to extract the open/flatten logic into a helper function where `defer` can be used efficiently.
```suggestion
```

kasabe28 commented:
As now we have another PR for source `Volume` and `VolumeSnapshot` deletion, which has changes from this PR also. I have addressed all comments in another PR -> https://github.com/ironcore-dev/ceph-provider/pull/813.
